File: backend/controllers/messaging_controllers.go
package controllers

import (
	"log"
	"time"

	"eunity.com/backend-main/helpers/WSManager"
	"eunity.com/backend-main/models"
	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

type Messaging_controllers struct {
}

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

// Send_message godoc
// @Summary Send a message
// @Description Send a message
// @Tags Messaging
// @Accept  json
// @Produce  json
// @Param message query string true "Message to send"
// @Param userID query string true "userId of the user"
// @Success 200 {object} string
// @Router /messaging/send [post]
func (m *Messaging_controllers) Send_message(c *gin.Context) {
	message := c.Query("message")
	//get user id from context
	from := c.GetString("user_id")
	to := c.Query("user_id")

	//get this users id

	//get conversation from backend and use it to create a message object
	conversationID := "placeholder"

	var messageObject *models.Message
	messageObject = messageObject.NewMessage(conversationID, from, to, message)

	//just echo message back
	err := WSManager.SendMessage(messageObject)
	if err != nil {
		c.JSON(500, gin.H{
			"message": "Error sending message",
		})
		return
	}

	c.JSON(200, gin.H{
		"success": true,
	})
}

func (m *Messaging_controllers) Web_Socket_Connection(c *gin.Context) {

	//get user id from context
	user_id := c.GetString("user_id")
	log.Println("User id: ", user_id)

	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Println(err)
		return
	}
	//add connection to connections
	WSManager.AddConnection(user_id, conn)

	defer conn.Close()
	for {
		time.Sleep(5 * time.Second)
	}
}

File: python/controllers/messaging_controllers.py
from fastapi import FastAPI, WebSocket, Query
from pydantic import BaseModel
import logging
import time
import asyncio

app = FastAPI()

logger = logging.getLogger(__name__)

ws_manager = None  # Initialize WSManager instance


class Message(BaseModel):
    conversation_id: str
    from_user: str
    to_user: str
    message: str


@app.post("/messaging/send")
async def send_message(
    message: str = Query(...), user_id: str = Query(...), to_user_id: str = Query(...)
):
    from_user_id = user_id
    conversation_id = "placeholder"  # Replace with actual conversation ID
    message_obj = Message(
        conversation_id=conversation_id,
        from_user=from_user_id,
        to_user=to_user_id,
        message=message,
    )
    try:
        await ws_manager.send_message(message_obj)
        return {"success": True}
    except Exception as e:
        return {"message": "Error sending message"}, 500


@app.websocket("/messaging/ws")
async def websocket_endpoint(websocket: WebSocket, user_id: str = Query(...)):
    logger.info(f"User ID: {user_id}")
    await websocket.accept()
    ws_manager.add_connection(user_id, websocket)
    while True:
        await websocket.send("Ping")
        await websocket.receive_text()  # just to keep the connection open
        await asyncio.sleep(5)
